File: fsharp/bank-account/BankAccount.fs
module BankAccount

open System

type Account = Account of decimal option ref * Object

let mkBankAccount() = Account(ref None, obj)

let update (Account(balance, lockObj) as account) f =
    lock lockObj (fun () -> f balance)
    account

let openAccount account = update account (fun b -> b.Value <- Some 0.0m)

let closeAccount account = update account (fun b -> b.Value <- None)

let getBalance (Account(balance, _)) = balance.Value

let updateBalance change account = update account (fun b -> b.Value <- b.Value |> Option.map ((+) change))

File: csharp/bank-account/BankAccount.cs
using System;

public class BankAccount
{
    private decimal _balance;
    private readonly object _balanceLock = new();

    private bool Active { get; set; }

    public void Open()
    {
        Active = true;
        _balance = 0;
    }

    public void Close() => Active = false;
    public decimal Balance => !Active ? throw new InvalidOperationException() : _balance;

    public void UpdateBalance(decimal change)
    {
        lock (_balanceLock) _balance += change;
    }
}
